subtests/docker_cli/deferred_deletion/deferred_deletion.py

```diff
@@ -0,0 +1,153 @@
+  3) dockerd unit file DOES NOT include MountFlags=slave; and
+  4) docker daemon is configured to use devicemapper storage; and
+  5) docker daemon is run with --storage-opts dm.use_deferred_deletion=true
+
+As a precondition for running this test, we check only conditions 4 & 5,
+aborting with TestNAError if either is false. We trust docker-storage-setup
+to only enable #5 if all preconditions are met. Anyone running a system
+with 4 & 5 true but 1, 2, or 3 false deserves what they get.
+
+Operational Summary
+----------------------
+
+#. Run a docker container
+#. Find out the local (host) mountpoint of its root filesystem. cd into it.
+#. Read the Deferred-Deletion count from 'docker info'.
+#. Remove the container.
+#. Read the Deferred-Deletion count; confirm that it has grown by one.
+#. cd back out of the container rootfs
+#. Read the Deferred-Deletion count; confirm that it has gone back down.
+
+"""
+
+import os
+from autotest.client import utils
+import dockertest.docker_daemon
+from dockertest import subtest
+from dockertest.containers import DockerContainers
+from dockertest.dockercmd import AsyncDockerCmd, DockerCmd
+from dockertest.images import DockerImage
+from dockertest.output import DockerInfo, mustpass
+from dockertest.xceptions import DockerTestNAError
+
+
+class deferred_deletion(subtest.Subtest):
+
+    def initialize(self):
+        super(deferred_deletion, self).initialize()
+        dockerinfo = DockerInfo()
+        storage_driver = dockerinfo['Storage Driver']
+        if storage_driver != 'devicemapper':
+            raise DockerTestNAError("This test is only applicable when using"
+                                    " the devicemapper storage driver; host"
+                                    " system is using %s" % storage_driver)
+
+        docker_cmdline = dockertest.docker_daemon.cmdline()
+        if "dm.use_deferred_deletion=true" not in str(docker_cmdline):
+            raise DockerTestNAError("deferred deletion not enabled in"
+                                    " docker daemon command-line options.")
+
+    def run_once(self):
+        super(deferred_deletion, self).run_once()
+
+        self._start_idle_container()
+
+        # cd into its directory. This prevents devicemapper from reaping it.
+        os.chdir(self._container_mountpoint)
+
+        # Determine baseline count of deferred deletions. We expect 0,
+        # but let's not panic if it isn't.
+        defer_count_initial = self._defer_count()
+        if defer_count_initial != 0:
+            self.logwarning("Deferred-Delete count from 'docker info'"
+                            " is %d (I expected 0). This may mean your"
+                            " system has still-undeleted containers.",
+                            defer_count_initial)
+        self.stuff['defer_count_initial'] = defer_count_initial
+
+        # Tell the container to exit
+        self._stop_idle_container()
```

Comment:
This should probably also go into ``cleanup()``, in case something above breaks.